### PR TITLE
[stable9] Backport lazy init of shared storage/mount

### PR DESCRIPTION
**Pull request by PVince81**
## Description
- Backport of https://github.com/owncloud/core/pull/23919/commits/2f1c62ce0b022a2549ff80a299b8a3b314106fae from #23919 to introduce `LazyStorageMountInfo`
- Backport of https://github.com/owncloud/core/pull/25789/commits/c8a385d61e3e534319defb94766d99a20456371e from #25789 to have a shortcut to retrieve the numeric storage id without having to initialize the shared storage fully
- Backport of https://github.com/owncloud/core/pull/25789/commits/dc7d55c857707360d70fa013701424d3696af8ce to have an additional safety against unwanted recursions in shared storages (could not make it happen on this PR but the extra safety won't harm)

## Related Issue
Hopefully fixes the issue described in https://github.com/owncloud/core/issues/26702

## Motivation and Context
There's a perf regression in 9.0 that didn't exist in 8.2 due to the introduction of oc_mounts.
Perf was improved in 9.1 through https://github.com/owncloud/core/pull/25789 so hopefully backporting a subset of it will also improve performance in 9.0.

## How Has This Been Tested?

Steps:

1. Setup OC 9.0
1. Run https://gist.github.com/PVince81/4c4c2e938ae29065dcf02cdc2cc6d0bc (requires latest pyocclient)
1. `curl -D - -X GET -u recipient:recipient http://localhost/owncloud/remote.php/webdav/file1.txt > file.txt` => ignore the result
1. `curl -D - -X GET -u recipient:recipient http://localhost/owncloud/remote.php/webdav/file1.txt > file.txt`, again, and have a look at the value of "Time spent"

Before this fix (v9.0.7): time spent was 7 seconds
After this fix: time spent is down to 4 seconds

Running the same test on v9.1.3 also gives 4 seconds.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## TODOs:

⚠️ it is likely that the introduction of this laziness will cause a regression due to a new recursive behavior (this was observed on 9.1, need to confirm if it will appear as well here through this PR). This was fixed by the PR https://github.com/owncloud/core/pull/25789 which itself brought its own regressions fixed by subsequent PRs.

- [x] check if there is a simpler shortcut => yes, but might not be worth it if we need additional backports anyway due to laziness (shortcut is on branch "stable9-shortcut-sharedmount-ids")
- [x] analyze whether this PR now contains the shared storage recursion hell and requires further backports => no recursion observed, but still did an extra safety backport
- [x] no further backports needed as this is not a `Jail` instance, so no further regression that were observed on 9.1 would happen
- [x] @mrow4a please rerun your powerful tests on this branch

### CODE REVIEW

**Comment by PVince81** [comment]
Good news: the `SharedStorage` in stable9 is not a Jail, so no need to backport the jail fix from https://github.com/owncloud/core/pull/25789/commits/3717e6237e10f42b6a73980d3033a6b5215e6bf8.

Now there is still a risk that the recursion happens when calling `$this->ownerView->getPath()`, I'll debug into it to see if it happens. If yes, we'll need https://github.com/owncloud/core/pull/25789/commits/dc7d55c857707360d70fa013701424d3696af8ce

- [ ] also backport the FailedStorage case, for more robustness ? https://github.com/owncloud/core/pull/25789/commits/837bf1c587b9ee55512390028cac9657cc0d81ea

**Comment by PVince81** [comment]
Attempting to find steps to reproduce the recursion in question, because we never really had steps for that... Here https://github.com/owncloud/core/issues/25557#issuecomment-271542853

**Comment by PVince81** [comment]
Could not reproduce the infinite recursion on this PR here https://github.com/owncloud/core/issues/25557#issuecomment-271549804.

It is likely that the code is too different. But there is still a slight chance that some code paths triggers it, so I decided to backport the recursion safety commit as well as it won't hurt other code paths: https://github.com/owncloud/core/pull/25789/commits/dc7d55c857707360d70fa013701424d3696af8ce

The FailedStorage backport looks unnecessary as the SharedStorage is not a `Jail` instance like on 9.1 where a lot of logic is different.

I think this PR is now ready for review and testing.

**Comment by PVince81** [comment]
@mrow4a this is ready for testing, please run your magic tests on this branch. Thanks

**Comment by PVince81** [comment]
Looks like https://github.com/owncloud/core/pull/26912/commits/e6d8779d70ba5bdea7473622242123c880684bba breaks some unit tests:
```
	Test Result (2 failures )

    Test_Files_Sharing_Api.testGetShareFromFolderReReShares
    Test_Files_Sharing_Api.testGetShareFromFileReReShares
```

**Comment by PVince81** [comment]
This because the tests create shares the old way which doesn't create flat reshares.
Tried to backport 6509220f824e906a5e7ffde27b4b334872852dce to have it use the proper APIs, but it makes the tests fail differently now.

**Comment by PVince81** [comment]
I had to compile a few backports to make tests pass again. The reason is that the getPath() shortcut is based on the assumption that reshares are flat. But the unit tests use the old API which uses the old chain-like shares. Note that the old API isn't used any more.

- [ ] REQUIRES: this other backport PR https://github.com/owncloud/core/pull/26927 to be merged first before tests can pass here.

**Comment by PVince81** [comment]
Looks like I'll have to figure out how to run the complex performance tests myself...

Can I at least get a code review ?

**Comment by PVince81** [comment]
Also, am running smashbox against this branch, just in case.

**Comment by mrow4a** [comment]
@PVince81 I have everything being set up, OC server with different versions, smashbox appliance etc. I just need to checkout the branch on one of server VM, put correct config files and run the tests on "smashbox" VM. 

Everything is there, I will try to finish quickly university stuff and help you -> https://malibu.int.owncloud.com:8006/ on internal network, password you know,

**Comment by PVince81** [comment]
@mrow4a thanks a lot!

File: lib/private/files/config/usermountcache.php [hunk 1]
```diff
@@ -91,18 +91,18 @@ public function registerMounts(IUser $user, array $mounts) {
 		});
 		/** @var ICachedMountInfo[] $newMounts */
 		$newMounts = array_map(function (IMountPoint $mount) use ($user) {
-			$storage = $mount->getStorage();
-			if ($storage->instanceOfStorage('\OC\Files\Storage\Shared')) {
-				$rootId = (int)$storage->getShare()['file_source'];
+			// note: SharedMount goes there
+			if (method_exists($mount, 'getStorageRootId')) {
```
Comment (PhilippSchaffrath) [inline lib/private/files/config/usermountcache.php]: Why is this necessary? `IMountPoint` has this method, if it does not exist you would either get an error for something not implementing `IMountPoint` correctly, or not injecting an object of type `IMountPoint`.

Comment (PVince81) [inline lib/private/files/config/usermountcache.php]: `IMountPoint` only has this method on stable9.1 or master and we don't want to introduce new methods as it qualifies as API change. Other implementers of IMountPoint (ex: external storage) will likely not implement this one in this version. So this hack is there to accomodate for that.

Comment (PhilippSchaffrath) [inline lib/private/files/config/usermountcache.php]: Ah, i didn't think about that! Makes sense then!

**Comment by PVince81** [comment]
Rebased to include https://github.com/owncloud/core/pull/26927, tests should pass now.

**Comment by PVince81** [comment]
@mrow4a has just confirmed that this improves performance significantly, and some results aren't exponential any more. The results aren't too close to master, but good enough to be released.

Please review the code @PhilippSchaffrath @jvillafanez 

This PR will also need some regression testing @owncloud/qa

File: lib/private/files/config/lazystoragemountinfo.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+	/**
+	 * CachedMountInfo constructor.
+	 *
+	 * @param IUser $user
+	 * @param IMountPoint $mount
+	 */
+	public function __construct(IUser $user, IMountPoint $mount) {
+		$this->user = $user;
+		$this->mount = $mount;
```
Comment (jvillafanez) [inline lib/private/files/config/lazystoragemountinfo.php]: I think we have to call the parent constructor. The parent might not be properly initialized otherwise and could cause issues.
If this is intended, write a comment explaining it.

File: lib/private/files/view.php [hunk 3]
```diff
@@ -1663,6 +1665,11 @@ public function getPath($id) {
 			/**
 			 * @var \OC\Files\Mount\MountPoint $mount
 			 */
+			if (!$includeShares && $mount instanceof SharedMount) {
```
Comment (jvillafanez) [inline lib/private/files/view.php]: I'm not sure if this will be enough to prevent the infinite loop

**Comment by jvillafanez** [comment]
I think we'll have to review the `lazystoragemountinfo.php` the sooner the better. It looks like a bad design, and this kind of things are always difficult to get rid of if they're left alone for a long time.

First I think it's pointless to have both classes (`CachedMountInfo` and `LazyStorageMountInfo`): you'll do it lazily or not, but you won't do it lazily sometimes and then normal.

Second, usually the class on top (`CachedMountInfo` in this case) is the one that has control over what the subclasses are allowed to do, and only a few methods are overwritten in the subclasses. Taking into account that we have control over both classes (`CachedMountInfo` and `LazyStorageMountInfo`), there is no reason not to modify the `CachedMountInfo` accordingly to allow lazyness in a clearer way, instead of overwrite almost all the methods and using protected parameters that aren't documented as "allowed to be used this way" in the subclasses or something similar.

I don't expect to solve this in this PR because it's a backport, but we have to take this into account.

File: lib/private/files/view.php (hunk 3 above)
Comment (PVince81) [inline lib/private/files/view.php]: This was enough on 9.1, at least for known cases

File: lib/private/files/config/lazystoragemountinfo.php (hunk 2 above)
Comment (PVince81) [inline lib/private/files/config/lazystoragemountinfo.php]: This was backported from 9.1 which didn't call the parent either (not my code), I hope it had a good reason

**Comment by PVince81** [comment]
> First I think it's pointless to have both classes (CachedMountInfo and LazyStorageMountInfo): you'll do it lazily or not, but you won't do it lazily sometimes and then normal.

You're right. I'm also not too happy about these. Would you suggest merging both classes into a single one ? (in a separate tech debt PR) That would likely solve all points you mentionned (calling parent, always lazy and no weird control)

**Comment by jvillafanez** [comment]
Users won't care about the specific implementation as long as it's fast enough, and a don't think there are cases where one implementation would perform better than the other one in order to provide a switch for those cases, so one implementation is enough.

File: apps/files_sharing/lib/sharedmount.php [hunk 4]
```diff
@@ -58,9 +61,10 @@ class SharedMount extends MountPoint implements MoveableMount {
 	public function __construct($storage, array $mountpoints, $arguments = null, $loader = null) {
 		$this->user = $arguments['user'];
 		$this->recipientView = new View('/' . $this->user . '/files');
-		$newMountPoint = $this->verifyMountPoint($arguments['share'], $mountpoints);
+		$this->share = $arguments['share'];
```
Comment (PVince81) [inline apps/files_sharing/lib/sharedmount.php]: fixed

**Comment by PVince81** [comment]
@jvillafanez I added the required checks. Your `LazyCacheMountInfo` can be adressed separately on master as tech debt: https://github.com/owncloud/core/issues/26960

**Comment by jvillafanez** [comment]
👍
